hack/lib/golang.sh

```diff
@@ -581,6 +581,9 @@ kube::golang::setup_env() {
   # Make sure our own Go binaries are in PATH.
   export PATH="${KUBE_GOPATH}/bin:${PATH}"
 
+  # In workspace mode, -mod flag may only be set to readonly or vendor.
+  export GOFLAGS=-mod=readonly
```

Comment:
Reading between the lines, it sounds like @tessapham may have GOFLAGS set (which is OK and we do want to respect), including -mod=mod (which is not OK).  That's the ony way I can reproduce the original report.  

We can either a) fail and let users change their calling env (most correct, IMO); b) unilaterally override GOFLAGS and set `-mod=vendor` (meaning you can't set -mod=readonly, which may be OK).

This seems to work (but would need comments, etc)

```diff
diff --git a/hack/lib/golang.sh b/hack/lib/golang.sh
index 1a65f82154b..dfc701951f4 100755
--- a/hack/lib/golang.sh
+++ b/hack/lib/golang.sh
@@ -589,6 +589,8 @@ kube::golang::setup_env() {
   unset GO111MODULE
   unset GOWORK
 
+  export GOFLAGS="${GOFLAGS:-} -mod=vendor"
+
   # This may try to download our specific Go version.  Do it last so it uses
   # the above-configured environment.
   kube::golang::internal::verify_go_version
```

Perhaps we should just do:

```diff
commit 63660ab0f20330af5adad795d5325f6c8b2ea984
Good "git" signature for thockin@google.com with ED25519 key SHA256:PfQ0rwNUgsu5aRmerT0vkihWn/S3MXY3uoCPUiMdPrg
Author: Tim Hockin <thockin@google.com>
Date:   Wed May 22 09:22:35 2024 -0700

    hack

diff --git a/hack/lib/golang.sh b/hack/lib/golang.sh
index dfc701951f4..1f44d579413 100755
--- a/hack/lib/golang.sh
+++ b/hack/lib/golang.sh
@@ -558,6 +558,12 @@ EOF
 #   env-var GOBIN is unset (we want binaries in a predictable place)
 #   env-var PATH includes the local GOPATH
 kube::golang::setup_env() {
+  # Sometimes users want to override Go build flags, but it's rare and can
+  # cause build failures (e.g. setting '-mod=mod' will not work).
+  if [[ -n "${GOFLAGS}" ]]; then
+      kube::log::info "WARNING: GOFLAGS is set, which may cause build issues: ${GOFLAGS}"
+  fi
+
   # Even in module mode, we need to set GOPATH for `go build` and `go install`
   # to work.  We build various tools (usually via `go install`) from a lot of
   # scripts.
```